### PR TITLE
Adding the ability to display subtags in the tagcloud

### PR DESCRIPTION
**Pull request by Lucas-C**
The counters in the tag cloud become clickable.
A demo of this feature can be seen here: https://chezsoi.org/shaarli/?do=tagcloud&searchtags=Python
If you like the idea, please tell me which tests I need to add to this PR ;)

### CODE REVIEW

File: application/LinkDB.php [hunk 1]
```diff
@@ -457,9 +457,19 @@ public function filterSearch($filterRequest = array(), $casesensitive = false, $
      */
     public function allTags()
     {
+        return $this->tagsOfLinksWithGivenTags();
```
Comment (ArthurHoaro) [inline application/LinkDB.php]: Just use one method if you only call the other one.

File: application/LinkDB.php [hunk 2]
```diff
@@ -457,9 +457,19 @@ public function filterSearch($filterRequest = array(), $casesensitive = false, $
...
+    }
+
+    /**
+     * Returns the list of tags appearing in the links matching the given tags
+     * Output: associative array key=tags, value=0
+     */
+    public function tagsOfLinksWithGivenTags($filteringTags = array(), $visibility = 'all')
```
Comment (ArthurHoaro) [inline application/LinkDB.php]: For new arrays, please use `[]` syntax.

File: application/LinkDB.php [hunk 3]
```diff
@@ -457,9 +457,19 @@ public function filterSearch($filterRequest = array(), $casesensitive = false, $
      */
     public function allTags()
     {
+        return $this->tagsOfLinksWithGivenTags();
+    }
+
+    /**
+     * Returns the list of tags appearing in the links matching the given tags
+     * Output: associative array key=tags, value=0
```
Comment (ArthurHoaro) [inline application/LinkDB.php]: For new comments, please use `@param` and `@return` PHPDoc tags.

File: tpl/default/tagcloud.html [hunk 4]
```diff
@@ -21,7 +21,7 @@ <h2 class="window-title">{'Tag cloud'|t} - {$countTags} {'tags'|t}</h2>
     <div id="cloudtag">
       {loop="tags"}
         <a href="?searchtags={$key|urlencode}" style="font-size:{$value.size}em;">{$key}</a
-        ><span class="count">{$value.count}</span>
+        ><a href="?addtag={$key|urlencode}" class="count">{$value.count}</a>
```
Comment (ArthurHoaro) [inline tpl/default/tagcloud.html]: Add a title `attribute` shortly explaining what it does.

File: application/LinkDB.php (hunk 1 above)
Comment (Lucas-C) [inline application/LinkDB.php]: Then should I replace every occurence of `allTags` by `tagsOfLinksWithGivenTags` ?

Comment (ArthurHoaro) [inline application/LinkDB.php]: I'd rather have the method renamed to `tags()`, but yes you'll also need to rename calls.

**Comment by Lucas-C** [comment]
Changes done

**Comment by Lucas-C** [comment]
I forgot something in terms of CSS.
In https://github.com/shaarli/Shaarli/blob/master/tpl/default/css/shaarli.css#L214 the `#search-tagcloud` ID should be added:
```
#search, #search-linklist, #search-tagcloud {
    text-align: center;
    width: 100%;
}
```
This is important for small devices (short width) : without the `text-align: center` it looks a little bit broken.
Should I send a new PR ?

**Comment by ArthurHoaro** [comment]
No you're fine, I didn't merge it yet.

**Comment by Lucas-C** [comment]
Great !
Fix pushed in my fork

**Comment by ArthurHoaro** [comment]
Thank you!

**Pull request merged**